### PR TITLE
[HttpKernel] Deprecated commands auto-registration

### PR DESCRIPTION
**Pull request by GuilhemN**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | no 
| BC breaks?    | no
| Deprecations? | yes 
| Tests pass?   | yes
| Fixed tickets | https://github.com/symfony/symfony/issues/23488
| License       | MIT
| Doc PR        | 

Deprecates commands auto-registration. See https://github.com/symfony/symfony/issues/23488 for arguments.

### CODE REVIEW

**Comment by GuilhemN** [comment]
Looks like we need to update the DoctrineBundle first.

File: UPGRADE-3.4.md [hunk 1]
```diff
@@ -114,6 +114,35 @@ FrameworkBundle
    class has been deprecated and will be removed in 4.0. Use the
    `Symfony\Component\Translation\DependencyInjection\TranslatorPass` class instead.
 
+HttpKernel
+----------
+
+ * Relying on commands auto-discovery has been deprecated and won't be supported
```
Comment (chalasr) [inline UPGRADE-3.4.md]: should we put `convention-based`?

File: UPGRADE-4.0.md [hunk 2]
```diff
@@ -452,6 +452,32 @@ HttpFoundation
 HttpKernel
 ----------
 
+ * Relying on commands auto-discovery has been deprecated and won't be supported
```
Comment (chalasr) [inline UPGRADE-4.0.md]: is not supported anymore

File: UPGRADE-3.4.md [hunk 3]
```diff
@@ -114,6 +114,35 @@ FrameworkBundle
...
+
+   ```yml
+   # app/config/services.yml
+   services:
+       # ...
+
+       # explicit commands registration
+       AppBundle\Command:
+           resource: '../../src/AppBundle/Command/*'
+           tags: ['console.command']
```
Comment (ogizanagi) [inline UPGRADE-3.4.md]: Not required when using auto configuration. Given the [default config](https://github.com/symfony/recipes/blob/master/symfony/framework-bundle/3.3/config/services.yaml#L16-L22) that'll be promoted, should we really add this before/after at all? Or at least mention it's not required when using the new default config?

Comment (GuilhemN) [inline UPGRADE-3.4.md]: This is for people upgrading from the standard edition < 3.3, the others won't get a deprecation so I don't think they will even notice the change.

Comment (ogizanagi) [inline UPGRADE-3.4.md]: Fair point 👍

File: UPGRADE-3.4.md (hunk 1 above)
Comment (GuilhemN) [inline UPGRADE-3.4.md]: is it better like that? :)

File: UPGRADE-4.0.md (hunk 2 above)
Comment (GuilhemN) [inline UPGRADE-4.0.md]: indeed

**Review by ogizanagi — APPROVED** [review]

**Review by chalasr — APPROVED** [review]

File: src/Symfony/Component/HttpKernel/Bundle/Bundle.php [hunk 4]
```diff
@@ -191,6 +191,8 @@ public function registerCommands(Application $application)
             }
             $r = new \ReflectionClass($class);
             if ($r->isSubclassOf('Symfony\\Component\\Console\\Command\\Command') && !$r->isAbstract() && !$r->getConstructor()->getNumberOfRequiredParameters()) {
+                @trigger_error(sprintf('Auto-registration of the command "%s" is deprecated since Symfony 3.4 and won\'t be supported in 4.0. Use PSR-4 based Service Discovery instead.', $class), E_USER_DEPRECATED);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Bundle/Bundle.php]: is using PSR4 registration enough? that won't remove that notice, isn't it? people need also to override the registerCommands method in their bundle?

Comment (GuilhemN) [inline src/Symfony/Component/HttpKernel/Bundle/Bundle.php]: No need as long as they register all their commands as services.

Comment (chalasr) [inline src/Symfony/Component/HttpKernel/Bundle/Bundle.php]: enough indeed, registering the command as a service make convention based skipped thanks to the `console.command.ids` parameter

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/HttpKernel/Bundle/Bundle.php (hunk 4 above)
Comment (Tobion) [inline src/Symfony/Component/HttpKernel/Bundle/Bundle.php]: I don't think "Service Discovery" should be written with capital letters as it doesn't reference a fixed term or feature name.

File: UPGRADE-4.0.md [hunk 5]
```diff
@@ -452,6 +452,32 @@ HttpFoundation
...
+   Use PSR-4 based Service Discovery instead.
+
+   Before:
+
+   ```yml
+   # app/config/services.yml
+   services:
+       # ...
+
+       # implicit commands registration
```
Comment (Tobion) [inline UPGRADE-4.0.md]: maybe add "implicit registration of all commands in the `Command` folder" for clarity

File: UPGRADE-3.4.md [hunk 6]
```diff
@@ -114,6 +114,35 @@ FrameworkBundle
    class has been deprecated and will be removed in 4.0. Use the
    `Symfony\Component\Translation\DependencyInjection\TranslatorPass` class instead.
 
+HttpKernel
+----------
+
+ * Relying on convention-based commands discovery has been deprecated and
+   won't be supported in 4.0. Use PSR-4 based service siscovery instead.
```
Comment (ogizanagi) [inline UPGRADE-3.4.md]: Typo :) (discovery)

Comment (GuilhemN) [inline UPGRADE-3.4.md]: Oops indeed :)

File: src/Symfony/Bundle/SecurityBundle/Tests/Functional/app/Acl/config.yml [hunk 7]
```diff
@@ -22,3 +22,9 @@ security:
                 users:
                     kevin: { password: test, roles: [ROLE_USER] }
                     anne:  { password: test, roles: [ROLE_ADMIN]}
+
+# to be removed once https://github.com/doctrine/DoctrineBundle/pull/684 is merged
+services:
+    Doctrine\Bundle\DoctrineBundle\Command\:
```
Comment (GuilhemN) [inline src/Symfony/Bundle/SecurityBundle/Tests/Functional/app/Acl/config.yml]: This is temporary :)

File: UPGRADE-3.4.md [hunk 8]
```diff
@@ -114,6 +114,35 @@ FrameworkBundle
...
+   won't be supported in 4.0. Use PSR-4 based service discovery instead.
+
+   Before:
+
+   ```yml
+   # app/config/services.yml
+   services:
+       # ...
+
+       # implicit commands registration
```
Comment (Tobion) [inline UPGRADE-3.4.md]: this should be updated as in the upgrade 4.0 file

Comment (GuilhemN) [inline UPGRADE-3.4.md]: done

**Review by Tobion — APPROVED** [review]

**Comment by Tobion** [comment]
Thank you @GuilhemN

**Pull request closed without merging**